src/flint/test/test_all.py

```diff
@@ -3093,19 +3146,29 @@ def quick_poly():
         # # XXX: Not sure what this should do in general:
         assert raises(lambda: pow(P(1, ctx=ctx), 2, 3), NotImplementedError)
 
-        if is_field:
-            assert (f * g).gcd(f) == f / 4
+        if (P is not flint.fmpz_mod_mpoly and P is not flint.nmod_mpoly) or f.context().is_prime():
+            if is_field:
+                assert (f * g).gcd(f) == f / 4
+            else:
+                assert (f * g).gcd(f) == f
+            assert raises(lambda: quick_poly().gcd(None), TypeError)
+            assert raises(lambda: quick_poly().gcd(P(ctx=ctx1)), IncompatibleContextError)
         else:
-            assert (f * g).gcd(f) == f
-        assert raises(lambda: quick_poly().gcd(None), TypeError)
-        assert raises(lambda: quick_poly().gcd(P(ctx=ctx1)), IncompatibleContextError)
+            assert raises(lambda: (f * g).gcd(f), DomainError)
 
-        assert (f * g).factor() == (S(2), [(mpoly({(0, 1): 1, (1, 0): 1}), 1), (f, 1)])
+        if P is flint.fmpz_mod_mpoly or P is flint.nmod_mpoly:
+            if is_field:
+                assert (f * g).factor() == (S(8), [(mpoly({(0, 1): 1, (1, 0): 1}), 1), (f / 4, 1)])
+        else:
+            assert (f * g).factor() == (S(2), [(mpoly({(0, 1): 1, (1, 0): 1}), 1), (f, 1)])
```

Comment:
If factor would raise an exception for non-prime modulus then that should be tested here so that the test shows the behaviour in all cases.

Also can these not be written with `g` in the output?
```python
        characteristic_zero = not (P is flint.fmpz_mod_mpoly or P is flint.nmod_mpoly)

        if characteristic_zero:
            # Primitive factors over Z for fmpz_mpoly and fmpq_mpoly
            assert (f * g).factor() == (S(2), [(g / 2, 1), (f, 1)])
        elif is_field:
            # Monic polynomials over Z/pZ for nmod_mpoly and fmpz_mod_mpoly
            assert (f * g).factor() == (S(8), [(g / 2, 1), (f / 4, 1)])
        else:
            # Factorisation not allowed over Z/nZ for n not prime.
            # Flint would abort so we raise an exception instead:
            assert raises(...)
```
One thing that I hadn't noticed until now is that `fmpq_mpoly.factor` behaves differently compared to `fmpq_poly.factor`. With `fmpq_poly` factors are always monic:
```python
In [29]: f = fmpq_poly([1, 4])

In [30]: g = fmpq_poly([2, 2])

In [31]: f
Out[31]: 4*x + 1

In [32]: g
Out[32]: 2*x + 2

In [33]: (f*g).factor()
Out[33]: (8, [(x + 1/4, 1), (x + 1, 1)])
```
However `fmpq_mpoly` gives factors primitive over Z:
```python
In [35]: ctx = fmpq_mpoly_ctx(1, Ordering.lex, ["x"])

In [36]: f = fmpq_mpoly('4*x + 1', ctx)

In [37]: g = fmpq_mpoly('2*x + 2', ctx)

In [38]: (f*g).factor()
Out[38]: (2, [(x + 1, 1), (4*x + 1, 1)])
```
I'm wondering whether it makes sense to try to make this consistent somehow. One sort of consistency is that whenever we have a field all factors can be monic which would suggest that fmpq_mpoly should also return monic factors.

On the other hand though it is computationally better for Q to use primitive factors and absorb any denominator into the constant factor as fmpq_mpoly does so we could change `fmpq_poly` to return primitive factors instead:
https://github.com/flintlib/python-flint/blob/ba1a7fe20ce8a4798896219c449b2ebc381c625a/src/flint/types/fmpq_poly.pyx#L396-L416
In SymPy it is `fmpq_poly` that is inconsistent with SymPy's behaviour and needs to be special-cased to remove precisely the denominators that python-flint inserts:
https://github.com/sympy/sympy/blob/31e957f47a636df9b7e62cd3951ec0afcc266369/sympy/polys/polyclasses.py#L2285-L2304

Reply:
> I'm wondering whether it makes sense to try to make this consistent somehow. One sort of consistency is that whenever we have a field all factors can be monic which would suggest that fmpq_mpoly should also return monic factors.
> 
> On the other hand though it is computationally better for Q to use primitive factors and absorb any denominator into the constant factor as fmpq_mpoly does so we could change fmpq_poly to return primitive factors instead:

In my mind I would expect python-flint to be largely transparent with the results it returns from flint, if flint is returning primitive factors then python-flint should as well, if not to just to be consistent. I've seen your #189 PR and will take a look now